File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 077485. 행렬 테두리 회전하기/kotlin/Solution.kt
class Solution {

    fun solution(rows: Int, columns: Int, queries: Array<IntArray>): IntArray {
        val board = Array(rows+1) {row -> IntArray(columns+1) { column -> ((columns) * (row-1)) + column }}
        fun rotate(sr:Int, sc:Int, er:Int, ec: Int): Int {
            var cr = sr
            var cc = sc
            fun nextRow(cr: Int, cc:Int): Int {
                return when {
                    (cr == sr && cc in sc until ec) -> cr
                    (cc == ec && cr in sr until er) -> cr + 1
                    (cr == er && cc in sc+1..ec) -> cr
                    else -> cr - 1
                }
            }
            fun nextColumn(cr: Int, cc: Int): Int {
                return when {
                    (cr == sr && cc in sc until ec) -> cc + 1
                    (cc == ec && cr in sr until er) -> cc
                    (cr == er && cc in sc + 1..ec) -> cc - 1
                    else -> cc
                }
            }
            var nr = nextRow(cr, cc)
            var nc = nextColumn(cr, cc)
            var backup = board[sr][sc]
            var result = backup
            while(nr != sr || nc != sc) {
                board[nr][nc] = backup.also{backup = board[nr][nc]}
                if (backup < result) result = backup
                cr = nr
                cc = nc
                nr = nextRow(cr, cc)
                nc = nextColumn(cr, cc)
            }
            board[sr][sc] = backup
            return result
        }
        val answer = IntArray(queries.size)
        for (i in queries.indices) {
            answer[i] = rotate(queries[i][0], queries[i][1], queries[i][2], queries[i][3])
        }
        return answer
    }


}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 077485. 행렬 테두리 회전하기/python/solution.py
def solution(rows, columns, queries):
    answer = []
    board = [[(r - 1) * columns + c for c in range(1, columns + 1)] for r in range(1, rows + 1)]

    def rotate(board, r1, c1, r2, c2):
        values = [board[r1][c1]]

        # 제일 왼쪽열 위로 밀기
        for r in range(r1 + 1, r2 + 1):
            values.append(board[r][c1])
            board[r - 1][c1] = board[r][c1]

        # 제일 아래 행 왼쪽으로 밀기
        for c in range(c1 + 1, c2 + 1):
            values.append(board[r2][c])
            board[r2][c - 1] = board[r2][c]

        # 제일 오른쪽 열 아래로 밀기
        for r in range(r2 - 1, r1 - 1, -1):
            values.append(board[r][c2])
            board[r + 1][c2] = board[r][c2]

        # 제일 위 열 오른쪽으로 밀기
        for c in range(c2 - 1, c1, -1):
            values.append(board[r1][c])
            board[r1][c + 1] = board[r1][c]

        board[r1][c1+1] = values[0]
        return min(values)

    for x1, y1, x2, y2 in queries:
        answer.append(rotate(board, x1 - 1, y1 - 1, x2 - 1, y2 - 1))
    return answer
